c10/cuda/CUDAFunctions.cpp

```diff
@@ -182,4 +182,130 @@ bool hasPrimaryContext(int64_t device_index) {
   return _internal::hasPrimaryContext(device_index);
 }
 
+// Wrappers for raw CUDA device management functions
+cudaError_t GetDeviceCount(int* dev_count) {
+  return cudaGetDeviceCount(dev_count);
+}
+
+// This is a codepath for CUDA 12 that comes with a critical change in behavior
+// of `cudaSetDevice`. Unlike to previous CUDA versions that allocate context
+// lazily CUDA 12.x eagerly allocates primary context the moment `cudaSetDevice`
+// is called. This can lead to dramatic consequences and pollute the device
+// memory in distributed runs. To avoid unnecessary context creation a new
+// function called `MaybeSetDevice` was introduced. This function is to be
+// called in device guard destructor and at the exit of torch.cuda.device
+// context manager. The behavior of `MaybeSetDevice` is quite simple, it calls
+// to `cudaSetDevice` if context already exist or if context was not allocated
+// on targeted device it simply saves the device index. This way we can keep
+// PyTorch backward compatible for applications like this:
+//
+// ```
+// import torch
+// x = torch.empty(1, device=“cuda:1”) # no CUDA context on cuda:0 after this
+// call y = torch.empty(1, device=“cuda”) # CUDA context is created on cuda:0
+// ```
+#if CUDA_VERSION >= 11000
+thread_local int targetDeviceIndex = -1;
+
+cudaError_t GetDevice(int* device) {
+  if (targetDeviceIndex >= 0) {
+    *device = targetDeviceIndex;
+    return cudaSuccess;
+  }
+  return cudaGetDevice(device);
+}
+
+cudaError_t SetDevice(int device) {
+  TORCH_CHECK(device >= 0, "device id must be positive!");
+  targetDeviceIndex = -1;
+  int cur_device = -1;
+  C10_CUDA_CHECK(cudaGetDevice(&cur_device));
+  if (device == cur_device) {
+    return cudaSuccess;
+  }
+  cudaError_t err = cudaSetDevice(device);
+  C10_CUDA_CHECK(cudaFree(0));
+  return err;
+}
+
+cudaError_t MaybeSetDevice(int device) {
+  if (hasPrimaryContext(device)) {
+    return c10::cuda::SetDevice(device);
+  }
+  targetDeviceIndex = device;
+  return cudaSuccess;
+}
+
+int ExchangeDevice(int to_device) {
+  int cur_device = targetDeviceIndex;
+  targetDeviceIndex = -1;
+  if (cur_device < 0) {
+    C10_CUDA_CHECK(cudaGetDevice(&cur_device));
+    if (to_device == cur_device) {
+      return cur_device;
+    }
+  }
+  C10_CUDA_CHECK(cudaSetDevice(to_device));
+  C10_CUDA_CHECK(cudaFree(0));
+  return cur_device;
+}
+
+int MaybeExchangeDevice(int to_device) {
```

Comment:
add comment that this function doesn't initialize the context on `to_device` if it doesn't exist already